In [6]:
!cd

In [7]:
from google.colab import files

uploaded = files.upload()

Saving the-verdict.txt to the-verdict (2).txt


In [8]:

with open("the-verdict.txt","r") as f:
    text_R = f.read()

In [9]:
print("Total number of character:", len(text_R))
print(text_R[:99])

Total number of character: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


In [10]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text)
print(result)
result = [item for item in result if item.strip() != ""]
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']
['Hello,', 'world.', 'This,', 'is', 'a', 'test.']


In [11]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text_R)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed)

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his', 'glory', ',', 'he', 'had', 'dropped', 'his', 'painting', ',', 'married', 'a', 'rich', 'widow', ',', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera', '.', '(', 'Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence', '.', ')', '"', 'The', 'height', 'of', 'his', 'glory', '"', '--', 'that', 'was', 'what', 'the', 'women', 'called', 'it', '.', 'I', 'can', 'hear', 'Mrs', '.', 'Gideon', 'Thwing', '--', 'his', 'last', 'Chicago', 'sitter', '--', 'deploring', 'his', 'unaccountable', 'abdication', '.', '"', 'Of', 'course', 'it', "'", 's', 'going', 'to', 'send', 'the', 'value', 'of', 'my', 'picture', "'", 'way', 'up', ';', 'but', 'I', 'don', "'", 't', 'think', 'of', 'that', ',

In [12]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1130


In [13]:
vocab = {token:integer for integer,token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i >= 50:
        break


('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)


In [14]:
from importlib.metadata import version
import tiktoken
print("tiktoken version:", version("tiktoken"))

tiktoken version: 0.12.0


In [15]:
tokenizer = tiktoken.get_encoding("gpt2")

text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of someunknownPlace."
)
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


In [16]:
print(len(text))
print(len(integers))

80
20


In [17]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


BPE tokenizer or byte-pair encoding tokenizer is based upon a sophisticated tokenization scheme which was used to train llms such as ChatGPT

It utilizes the <| endoftext |> token to join two different segments of text

the BPE tokenizer used to train GPT-2,3 and the original model in ChatGPT had a vocabulary size of 50,257 with <| endoftext |> being assigned the largest token id of 50,256

It utilizes an algorithm which breaks down the unknown words not in its vocabulary into a sequence of smaller sub-words or even individual characters to encode it.

Example:


In [18]:
integers = tokenizer.encode("Kabjngvf :;wfwg")
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[42, 397, 73, 782, 85, 69, 1058, 26, 86, 44482, 70]
Kabjngvf :;wfwg


As we see the "unknown" string was broken down by the tokenizer into a sequence of smaller segments which were effectively decoded too later on

In [19]:
import tiktoken
tokenizer = tiktoken.get_encoding("gpt2")
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [20]:
enc_sample = enc_text[50:]

context_size = 5
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y:      {y}")

x: [290, 4920, 2241, 287, 257]
y:      [4920, 2241, 287, 257, 4489]


In [21]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, "---->", desired)

[290] ----> 4920
[290, 4920] ----> 2241
[290, 4920, 2241] ----> 287
[290, 4920, 2241, 287] ----> 257
[290, 4920, 2241, 287, 257] ----> 4489


In [22]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

 and ---->  established
 and established ---->  himself
 and established himself ---->  in
 and established himself in ---->  a
 and established himself in a ---->  vill


In [23]:
from torch.utils.data import Dataset, DataLoader

In [24]:
import tiktoken
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
tokenizer = tiktoken.get_encoding("gpt2")

In [25]:
import torch

In [26]:
class GPTDataset(Dataset):
    def __init__(self, text, tokenizer, context_size, stride):
        self.input_ids = []
        self.target_ids = []
        tokenized_text = tokenizer.encode(text)
        for i in range(0, len(tokenized_text) - context_size, stride):
            input_id = tokenized_text[i:i+context_size]
            target_id = tokenized_text[i+1:i+context_size+1]
            self.input_ids.append(input_id)
            self.target_ids.append(target_id)

    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, sample_id):
        return self.input_ids[sample_id], self.target_ids[sample_id]

In [27]:
dataset = GPTDataset(
    raw_text, tokenizer,
    context_size=4,
    stride=1
)

In [28]:
dataset[0]

([40, 367, 2885, 1464], [367, 2885, 1464, 1807])

In [29]:
len(dataset)

5141

In [30]:
raw_text[:100]

'I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no g'

In [31]:
tokenizer.decode(dataset[0][0])
tokenizer.decode(dataset[0][1])

' HAD always thought'

In [32]:
def data_loader(text, batch_size, context_window = 256, stride = 64, shuffle = False,num_workers = 3,drop_last = True):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset = GPTDataset(
        text, tokenizer,
        context_size=context_window,
        stride=stride
    )
    return DataLoader(dataset, batch_size=batch_size, shuffle=shuffle, num_workers=num_workers, drop_last=drop_last)


In [33]:
dataloader = data_loader(raw_text, batch_size=4, context_window = 256, stride = 64, shuffle=True, num_workers=3)

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()


In [34]:
x, y = dataloader.dataset[0]
print(tokenizer.decode(x))
print(tokenizer.decode(y))

I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no great surprise to me to hear that, in the height of his glory, he had dropped his painting, married a rich widow, and established himself in a villa on the Riviera. (Though I rather thought it would have been Rome or Florence.)

"The height of his glory"--that was what the women called it. I can hear Mrs. Gideon Thwing--his last Chicago sitter--deploring his unaccountable abdication. "Of course it's going to send the value of my picture 'way up; but I don't think of that, Mr. Rickham--the loss to Arrt is all I think of." The word, on Mrs. Thwing's lips, multiplied its _rs_ as though they were reflected in an endless vista of mirrors. And it was not only the Mrs. Thwings who mourned. Had not the exquisite Hermia Croft, at the last Grafton Gallery show, stopped me before Gisburn's "Moon-dancers" to say, with tears in her eyes: "We shall not look upon its like again"?

Well!--
 HAD always th

In [35]:
for i in range(len(dataloader.dataset)):

    x, y = dataloader.dataset[i]

    print(f"SAMPLE {i}")

    print("INPUT : ", tokenizer.decode(x))
    print("TARGET:", tokenizer.decode(y))

    print("-" * 50)

SAMPLE 0
INPUT :  I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no great surprise to me to hear that, in the height of his glory, he had dropped his painting, married a rich widow, and established himself in a villa on the Riviera. (Though I rather thought it would have been Rome or Florence.)

"The height of his glory"--that was what the women called it. I can hear Mrs. Gideon Thwing--his last Chicago sitter--deploring his unaccountable abdication. "Of course it's going to send the value of my picture 'way up; but I don't think of that, Mr. Rickham--the loss to Arrt is all I think of." The word, on Mrs. Thwing's lips, multiplied its _rs_ as though they were reflected in an endless vista of mirrors. And it was not only the Mrs. Thwings who mourned. Had not the exquisite Hermia Croft, at the last Grafton Gallery show, stopped me before Gisburn's "Moon-dancers" to say, with tears in her eyes: "We shall not look upon its like again"?

Well

In [38]:
for batch_x, batch_y in dataloader:

    print(batch_x)

    print(batch_y)

    break

[tensor([ 40, 517, 991, 314]), tensor([ 367,   13, 2045, 2138]), tensor([2885,  383,  546, 1807]), tensor([1464, 3200,  329,  340]), tensor([1807,   30,  257,  561]), tensor([ 3619,  4162, 12854,   423]), tensor([402,  11, 286, 587]), tensor([  271,   314,   884, 10598]), tensor([10899,   550,  3842,   393]), tensor([ 2138,   257,    13, 28537]), tensor([ 257, 3200,  198, 2014]), tensor([7026, 2861,  198,  198]), tensor([15632,  8208,     1,   198]), tensor([  438,   286, 12295,     1]), tensor([2016,  465,  553,  464]), tensor([ 257,    0,  339, 6001]), tensor([922, 314, 531, 286]), tensor([ 5891, 37901, 11589,   465]), tensor([ 1576,   379,    13, 13476]), tensor([438, 262, 198,   1]), tensor([  568, 21978,   198,   438]), tensor([  340, 44896,     1,  5562]), tensor([ 373,   11, 5574,  373]), tensor([ 645,  290, 1660,  644]), tensor([1049, 3088,   12,  262]), tensor([ 5975,   617, 49903,  1466]), tensor([ 284,  286,  438, 1444]), tensor([502, 616, 273, 340]), tensor([  284, 49025,  